sql/range_test.go

```diff
@@ -315,7 +315,96 @@ func TestRangeOverlapThreeColumns(t *testing.T) {
 	}
 }
 
-func setup() (x, y, z sql.Expression, values2, values3 [][]interface{}) {
+func TestRangeOverlapNulls(t *testing.T) {
+	ctx := sql.NewEmptyContext()
+	x, y, _, _, _, valuesNull := setup()
+
+	tests := []struct {
+		reference sql.Expression
+		ranges    sql.RangeCollection
+	}{
+		{
+			reference: or(
```

Comment:
Perhaps some more variation. The `or` is unnecessary as there's only one `and` within each, so the `or` can be removed from those tests. Then there should be more tests mixing multiple `and`s together inside of an `or`, or even multiple `or` and `and` combinations.

Reply:
Do you think the enginetests do our null patterns justice? I hit regular and multicol indexes with NULL/IS NOT NULL for the variety of filter types, which i think is all the code paths. The above unit tests were really hard to customize imo.

Reply:
I don't think our enginetests do null patterns justice even remotely. Non-null is pretty well tested, but I don't think the same care was given to null.

For the unit tests, they should be relatively straightforward. They essentially mimic the filter you'd write. For example:
```
or(
	and(lt(x, 2), gt(y, 5)),
	and(gt(x, 8), gt(y, 5)),
	and(gt(x, 5), gt(y, 8)),
)
```
is equivalent to
```
(x < 2 AND y > 5) OR (x > 8 AND y > 5) OR (x > 5 AND y > 8)
```
It's just converting from infix to prefix. Then to get the expected filters, you cut out the `or` since multiple filters implicitly mean `OR`. Then you figure out which filters the above would reduce down to, and write those basically the same as before. `r()` is a range filter, where each parameter corresponds to the column index (so the first is for `x`, second is for `y` in the above example). `rlt` is the same as `lt` just for ranges, and without the column since it's in the correct index position. And that's really it. So the above becomes:
```
sql.RangeCollection{
	r(rlt(2), rgt(5)),
	r(rgt(8), rgt(5)),
	r(rgt(5), rgt(8)),
},
```
Hopefully that explanation helps!

Reply:
I mean that there is an implicit required structure in the test expressions that requires a concatenation of 1 x expression and 1 y expression. That made it hard to test ex: only 1 x, or 2 x's, which i did add in my new enginetests. Our previous enginetests didn't touch nulls and were insufficient, but hopefully the new ones i added fill that gap?